src/raftstore/store/store.rs

```diff
@@ -1536,104 +1536,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+        if is_leader {
+            // Notify pd immediately to let it update the region meta.
+            if let Err(e) = report_split_pd(&regions, &self.pd_worker) {
+                error!("{} failed to notify pd: {}", self.tag, e);
             }
         }
-
-        let mut campaigned = false;
-        let peer;
-        match Peer::create(self, &new_region) {
-            Err(e) => {
-                // peer information is already written into db, can't recover.
-                // there is probably a bug.
-                panic!("create new split region {:?} err {:?}", new_region, e);
-            }
-            Ok(mut new_peer) => {
-                for peer in new_region.get_peers() {
-                    // Add this peer to cache.
-                    new_peer.insert_peer_cache(peer.clone());
-                }
-                peer = new_peer.peer.clone();
-                if let Some(origin_peer) = self.region_peers.get_mut(&region_id) {
-                    // New peer derive write flow from parent region,
-                    // this will be used by balance write flow.
-                    new_peer.peer_stat = origin_peer.peer_stat.clone();
-
-                    campaigned =
-                        new_peer.maybe_campaign(origin_peer, &mut self.pending_raft_groups);
-
-                    if origin_peer.is_leader() {
-                        // Notify pd immediately to let it update the region meta.
-                        if let Err(e) = if right_derive {
-                            report_split_pd(&mut new_peer, origin_peer, &self.pd_worker)
-                        } else {
-                            report_split_pd(origin_peer, &mut new_peer, &self.pd_worker)
-                        } {
-                            error!("{} failed to notify pd: {}", self.tag, e);
-                        }
-                    }
-                }
-
+        let last_key = enc_end_key(regions.last().unwrap());
+        self.region_ranges
+            .remove(&last_key)
+            .expect("original region should exists");
+        let last_region_id = regions.last().unwrap().get_id();
+        for new_region in regions {
+            let new_region_id = new_region.get_id();
+            if new_region_id != region_id {
                 // Insert new regions and validation
-                info!("insert new regions left: {:?}, right:{:?}", left, right);
-                if self.region_ranges
-                    .insert(enc_end_key(&left), left.get_id())
-                    .is_some()
-                {
-                    panic!("region should not exist, {:?}", left);
-                }
-                if self.region_ranges
-                    .insert(enc_end_key(&right), right.get_id())
-                    .is_none()
-                {
-                    panic!("region should exist, {:?}", right);
+                info!("insert new region {:?}", new_region);
```

Comment:
Could we move this line before L1581 exactly before initializing new peers and validation for the new region?